### ISSUE TITLE
java.util.ConcurrentModificationException in formatter

### ISSUE THREAD

**Issue by sigiesec** [eclipse/xtext-xtend#478]
When trying to format
```
package com.btc.serviceidl.generator.java

import com.btc.serviceidl.idl.AbstractTypeDeclaration
import com.btc.serviceidl.idl.AliasDeclaration
import com.btc.serviceidl.idl.ModuleDeclaration

class JavaGenerator
{
   private def void generateCommon(String projectSourceRootPath, ModuleDeclaration module)
   {
      for ( element : module.moduleComponents.filter(AbstractTypeDeclaration).filter[e | !(e instanceof AliasDeclaration)] )
      {
      }
      
   }   
}
```
the following exception is thrown and logged in the Eclipse Error Log:
```
java.util.ConcurrentModificationException
	at java.base/java.util.ArrayList$Itr.checkForComodification(Unknown Source)
	at java.base/java.util.ArrayList$Itr.next(Unknown Source)
	at com.google.common.collect.Iterators$3.next(Iterators.java:174)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.createReplacements(FormattableDocument.java:130)
	at org.eclipse.xtext.formatting2.internal.SubDocument.createReplacements(SubDocument.java:38)
	at org.eclipse.xtext.formatting2.internal.ConditionalReplacer.createReplacements(ConditionalReplacer.java:43)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.createReplacements(FormattableDocument.java:146)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.renderToTextReplacements(FormattableDocument.java:293)
	at org.eclipse.xtext.formatting2.AbstractFormatter2.format(AbstractFormatter2.java:275)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter.exec(ContentFormatter.java:90)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter$FormattingUnitOfWork.exec(ContentFormatter.java:57)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter$FormattingUnitOfWork.exec(ContentFormatter.java:1)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:520)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.priorityReadOnly(XtextDocument.java:485)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.priorityReadOnly(XtextDocument.java:142)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter.format(ContentFormatter.java:103)
	at org.eclipse.jface.text.source.SourceViewer.doOperation(SourceViewer.java:992)
	at org.eclipse.jface.text.source.projection.ProjectionViewer.doOperation(ProjectionViewer.java:1461)
	at org.eclipse.xtend.ide.editor.RichStringAwareSourceViewer.doOperation(RichStringAwareSourceViewer.java:260)
	at org.eclipse.ui.texteditor.TextOperationAction$1.run(TextOperationAction.java:130)
	at org.eclipse.swt.custom.BusyIndicator.showWhile(BusyIndicator.java:70)
	at org.eclipse.ui.texteditor.TextOperationAction.run(TextOperationAction.java:127)
	at org.eclipse.jface.action.Action.runWithEvent(Action.java:473)
	at org.eclipse.jface.commands.ActionHandler.execute(ActionHandler.java:118)
	at org.eclipse.ui.internal.handlers.E4HandlerProxy.execute(E4HandlerProxy.java:92)
	at jdk.internal.reflect.GeneratedMethodAccessor205.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.e4.core.internal.di.MethodRequestor.execute(MethodRequestor.java:55)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invokeUsingClass(InjectorImpl.java:305)
	at org.eclipse.e4.core.internal.di.InjectorImpl.invoke(InjectorImpl.java:239)
	at org.eclipse.e4.core.contexts.ContextInjectionFactory.invoke(ContextInjectionFactory.java:132)
	at org.eclipse.e4.core.commands.internal.HandlerServiceHandler.execute(HandlerServiceHandler.java:152)
	at org.eclipse.core.commands.Command.executeWithChecks(Command.java:494)
	at org.eclipse.core.commands.ParameterizedCommand.executeWithChecks(ParameterizedCommand.java:487)
	at org.eclipse.e4.core.commands.internal.HandlerServiceImpl.executeHandler(HandlerServiceImpl.java:210)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.executeCommand(KeyBindingDispatcher.java:287)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.press(KeyBindingDispatcher.java:527)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.processKeyEvent(KeyBindingDispatcher.java:577)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.filterKeySequenceBindings(KeyBindingDispatcher.java:385)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher.access$0(KeyBindingDispatcher.java:331)
	at org.eclipse.e4.ui.bindings.keys.KeyBindingDispatcher$KeyDownFilter.handleEvent(KeyBindingDispatcher.java:88)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:86)
	at org.eclipse.swt.widgets.Display.filterEvent(Display.java:1271)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1078)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1103)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1088)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1130)
	at org.eclipse.swt.widgets.Widget.sendKeyEvent(Widget.java:1126)
	at org.eclipse.swt.widgets.Widget.wmChar(Widget.java:1547)
	at org.eclipse.swt.widgets.Control.WM_CHAR(Control.java:4962)
	at org.eclipse.swt.widgets.Canvas.WM_CHAR(Canvas.java:364)
	at org.eclipse.swt.widgets.Control.windowProc(Control.java:4843)
	at org.eclipse.swt.widgets.Canvas.windowProc(Canvas.java:359)
	at org.eclipse.swt.widgets.Display.windowProc(Display.java:5178)
	at org.eclipse.swt.internal.win32.OS.DispatchMessageW(Native Method)
	at org.eclipse.swt.internal.win32.OS.DispatchMessage(OS.java:2560)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3815)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1150)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1039)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:153)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:680)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:336)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:594)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:148)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:151)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:196)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:134)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:388)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:243)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:653)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:590)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1499)
	at org.eclipse.equinox.launcher.Main.main(Main.java:1472)
```

**Comment by cdietrich** [comment]
is this a duplicate of https://github.com/eclipse/xtext-core/issues/527

**Comment by cdietrich** [comment]
additional problem 

```
1    [main] ERROR org.eclipse.xtext.util.ExceptionAcceptor  - 
org.eclipse.xtext.formatting2.FormattingNotApplicableException
	at org.eclipse.xtext.formatting2.internal.MaxLineWidthDocument.validate(MaxLineWidthDocument.java:55)
	at org.eclipse.xtext.formatting2.internal.MaxLineWidthDocument.validate(MaxLineWidthDocument.java:60)
	at org.eclipse.xtext.formatting2.internal.MaxLineWidthDocument.addReplacer(MaxLineWidthDocument.java:32)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.set(FormattableDocument.java:321)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.append(FormattableDocument.java:87)
	at org.eclipse.xtext.xbase.formatting2.XbaseFormatter.formatExpressionsMultiline(XbaseFormatter.java:1324)
	at org.eclipse.xtext.xbase.formatting2.XbaseFormatter._format(XbaseFormatter.java:1147)
	at org.eclipse.xtend.core.formatting2.XtendFormatter.format(XtendFormatter.java:702)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.format(FormattableDocument.java:188)
	at org.eclipse.xtext.xbase.formatting2.XbaseFormatter.lambda$19(XbaseFormatter.java:294)
	at org.eclipse.xtext.formatting2.internal.ConditionalReplacer.createReplacements(ConditionalReplacer.java:41)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.createReplacements(FormattableDocument.java:146)
	at org.eclipse.xtext.formatting2.internal.FormattableDocument.renderToTextReplacements(FormattableDocument.java:293)
	at org.eclipse.xtext.formatting2.AbstractFormatter2.format(AbstractFormatter2.java:280)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter.exec(ContentFormatter.java:90)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter$FormattingUnitOfWork.exec(ContentFormatter.java:57)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter$FormattingUnitOfWork.exec(ContentFormatter.java:1)
	at org.eclipse.xtext.resource.OutdatedStateManager.exec(OutdatedStateManager.java:91)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.internalReadOnly(XtextDocument.java:527)
	at org.eclipse.xtext.ui.editor.model.XtextDocument$XtextDocumentLocker.priorityReadOnly(XtextDocument.java:491)
	at org.eclipse.xtext.ui.editor.model.XtextDocument.priorityReadOnly(XtextDocument.java:146)
	at org.eclipse.xtext.ui.editor.formatting2.ContentFormatter.format(ContentFormatter.java:103)
	at org.eclipse.jface.text.source.SourceViewer.doOperation(SourceViewer.java:1031)
	at org.eclipse.jface.text.source.projection.ProjectionViewer.doOperation(ProjectionViewer.java:1461)
	at org.eclipse.xtend.ide.editor.RichStringAwareSourceViewer.doOperation(RichStringAwareSourceViewer.java:260)
	at org.eclipse.ui.texteditor.TextOperationAction$1.run(TextOperationAction.java:130)
```

**Comment by cdietrich** [comment]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=474468

**Comment by cdietrich** [comment]
BlankLineKey requires at least one newline.
but the validator forbids that
how can that ever work
@meysholdt any hints?

**Comment by cdietrich** [comment]
=> do we have to change the key as follows

```
class BlankLineKey extends IntegerKey implements Procedures.Procedure1<IHiddenRegionFormatter> {
	new(String name, Integer defaultValue) {
		super(name, defaultValue)
	}

	override apply(IHiddenRegionFormatter it) {
		val preferences = request.preferences
		val blankline = preferences.getPreference(this)
		val preserve = preferences.getPreference(preserveBlankLines)
		val min = blankline + 1
		val max = Math.max(preserve + 1, min)
		setNewLines(0, min, max)
	}

}
```

**Comment by cdietrich** [comment]
@szarnekow the validation is by you. maybe you have a hint too.

**Comment by szarnekow** [comment]
About the validation: Looks like the condition is plain wrong `if (newLineMin != null && newLineMin > 0)` should likely read `if (newLineMin != null && newLineMin < 0)`

**Comment by cdietrich** [comment]
Fixed in 2.15
